add glm support + patch (#3329)
OpenAccess-AI-Collective/axolotl: examples/glm46v/README.md

# Finetune GLM-4.6V with Axolotl

GLM-4.6V is a family of vision-language models from ZhipuAI found on [HuggingFace](https://huggingface.co/zai-org/GLM-4.6V). This guide shows how to fine-tune it with Axolotl for vision-language tasks.

## Getting started

1. Install Axolotl from source following the [installation guide](https://docs.axolotl.ai/docs/installation.html#sec-edge-build).

2. Install [Cut Cross Entropy](https://docs.axolotl.ai/docs/custom_integrations.html#cut-cross-entropy) to reduce training VRAM usage.

3. Swap to the Axolotl transformers v5 branch

    ```bash
    git fetch
    git checkout transformers-v5

    # Install packages for transformers v5
    pip install -e .
    ```

4. Run the fine-tuning:

    ```bash
    axolotl train examples/glm46v/glm-4-6v-qlora.yaml
    ```

    Or for the Flash variant:

    ```bash
    axolotl train examples/glm46v/glm-4-6v-flash-qlora.yaml

>     ```

Reviewer: Let's put the Flash one first as it's lighter. Maybe add a comment about the size of the model and  the Flash vram usage, put it there. See the gemma3n as example doc.

Let us know how it goes. Happy finetuning! 🚀

## Tips

- The configs use `processor_type: AutoProcessor` to automatically handle vision inputs.
- Vision datasets should follow the OpenAI Messages format with image content. See [multimodal docs](https://docs.axolotl.ai/docs/multimodal.html#dataset-format).
- You can run a full finetuning by removing the `adapter: qlora` and `load_in_4bit: true` from the config.
- Read more on how to load your own dataset at [docs](https://docs.axolotl.ai/docs/dataset_loading.html).

> - The dataset format follows the OpenAI Messages format as seen [here](https://docs.axolotl.ai/docs/dataset-formats/conversation.html#chat_template).

Reviewer: Point to the Vision data format in the multi-modal docs. See ministral3 /vision as example

Maybe you want to put a text-only version for train as well. To do so, just remove the vision config and train like a dense layer. You may need to set `model_type: GLM...` text class (if possible).

# Finetune GLM-4.6V with Axolotl

GLM-4.6V is a family of vision-language models from ZhipuAI found on [HuggingFace](https://huggingface.co/zai-org/GLM-4.6V). This guide shows how to fine-tune it with Axolotl for vision-language tasks.

## Getting started

1. Install Axolotl from source following the [installation guide](https://docs.axolotl.ai/docs/installation.html#sec-edge-build).

2. Install [Cut Cross Entropy](https://docs.axolotl.ai/docs/custom_integrations.html#cut-cross-entropy) to reduce training VRAM usage.

3. Swap to the Axolotl transformers v5 branch

    ```bash
    git fetch
    git checkout transformers-v5

    # Install packages for transformers v5
    pip install -e .

>     ```

Reviewer: Not required anymore, given we merge v5 in

4. Run the fine-tuning:

    glm-4-6v-flash(9B)
    ```bash
    axolotl train examples/glm46v/glm-4-6v-flash-qlora.yaml
    ```

Let us know how it goes. Happy finetuning! 🚀

## Tips

> - The configs use **`processor_type: AutoProcessor`** to automatically handle vision inputs.

Reviewer: ```suggestion
```

> - Vision datasets should follow the **vision data format** described in the [multimodal docs (Vision section)](https://docs.axolotl.ai/docs/multimodal.html#vision-data-format), similar to the `ministral3/vision` examples.

Reviewer: ```suggestion
- Vision datasets should follow the format described in the [multimodal docs](https://docs.axolotl.ai/docs/multimodal.html#dataset-format)
```

> - The dataset format is based on the **OpenAI Messages** schema with image content – see the [conversation format docs](https://docs.axolotl.ai/docs/dataset-formats/conversation.html#chat_template) for details.

Reviewer: ```suggestion
```
incorrect

- You can run a **full finetuning** by removing the `adapter: qlora` and `load_in_4bit: true` from the config.
- Read more on how to load your own dataset in the [dataset loading docs](https://docs.axolotl.ai/docs/dataset_loading.html).

## Text-only training (no vision)

- If you only want to finetune **text**:
  - Start from the same config and **remove the vision-specific fields** (e.g. `is_multimodal`, `image_column`, `image_size`, and any vision processor settings).
  - Train it like a standard dense LLM (similar to other text-only configs).

>   - Depending on the GLM checkpoints you use, you may need to set `model_type` to the appropriate **GLM text class** (e.g. the text-only GLM variant for that family), if auto-detection does not pick it up correctly.

Reviewer: I think this needs to be checked. Either provide a text config that folks can use or omit this section
